### ISSUE TITLE
Thoughts on switching

### ISSUE THREAD

**Issue by rikkertkoppes** [poelstra/ts-stream#21]
What are your thoughts on switching a stream? I.e. choosing different paths based on some test function.

This could be achieved by splitting (duplicating) stream and then filtering each, but as you mention, duplicating a stream is a bit of a thing. And not needed here, because any packet would only be allowed to end up into one child stream.

An 'end' event would need to propagate to all children though
Backpressure would be distributed over children I think, when one child stalls, the other would also stop receiving packets, since the write in the switch is blocked.
Errors would be distributed as well.

Basically, at any point in time, there would be only on "active" child, so any reasoning can be applied to a linear stream. This would probably not be the fastest way, because it does not really parallelize streams.

What are your thoughts on this? Does it fit the ts-stream roadmap?

What I am looking for is a good foundation to build a system to handle file transformations, like starting with a directory name, create a stream of filenames using glob, mapping that to a stream of files (read:pdf), further splitting that out into a stream of pages (for example starting with a header, then all pages and a pdf end packet), performing some operations on each page (cropping, inserting pages, merging pages), assembling it back together to a stream of pdf files, saving it, mapping the stream to file names, and eventually reducing the stream to an array of the result files.

This all would fit nicely in something like bacon, except it is lacking backpressure.

**Comment by rogierschouten** [comment]
I've created a standard splitter and then used standard filtering on the
output streams. It's not that much overhead in my case and it's reusable.
If necessary it would be easy to make a splitter that already includes the
filtering. Is that what you mean?

**Comment by rikkertkoppes** [comment]
Kinda, but also multiple (more than two) outputs

Also, I wonder how your splitter behaves if one output applies some backpressure, what would happen to the stream on the other ouput? Or the write?

**Comment by rogierschouten** [comment]
My stream splitter supports any number of outputs. It simply sends all writes to all outputs and waits until all promises are resolved before it returns its own result. Errors are collected an aggregated.
Similarly, the end() call is fowarded to all outputs in parallel and only when all have resolved their promises, the end() promise is resolved. Again, any errors are aggregated into one error.

In other words, if the source supports backpressure, it would wait after every write until all outputs are done, and then perform the next write. It can also decide not to wait and simply continue writing, which will result in multiple writes on the outputs. It's completely transparent. Any other approach would need buffering inside the splitter, and it also would make the splitter less generic. With the current approach you can decide outside of the splitter how to deal with backpressure.

**Comment by poelstra** [comment]
I agree with Rogier's comments, things like buffering and how to handle the back pressure should be explicit.

It could be nice to have an optional filter to a splitted stream to make it slightly more efficient, but if you're going to be reading and writing (pdf) files, I don't think this is going to be the bottleneck ;)

It's a use case that should fit very well, though!

**Comment by poelstra** [comment]
@rikkertkoppes Do the comments above answer your questions? If so, shall we close this issue?

Note: there already is an issue to implement a generic stream splitter, which would form the basis of a solution for your case I think.

**Comment by rikkertkoppes** [comment]
Yeah, ok for now. I am actively using ts-stream now (in a pdf page juggling
workflow with scissors (search github) and globs and file handling). More
questions and use cases may show up in the future. I am happy to think
along and contribute.

**Comment by poelstra** [comment]
Nice to hear, thanks!
